### PR TITLE
feat(monitor): Retry Logic

### PR DESCRIPTION
**Pull request by chrismaree**
Add retry logic to the monitor bot

**Motivation**

Right now the monitor bot has no notion of re-try logic. This makes it fail ungracefully in situations where a simple re-try would resolve the issue.

**Summary**

This PR implements the same logic added in PR [1838](https://github.com/UMAprotocol/protocol/pull/1838) for the liquidator bot. 

**Details**

One difference in this PRs implementation is how it is tested. This proved to be quite challenging as we've designed the monitor bot to _never_ throw errors in the main while loop. For example, invalid price feeds don't "error", they simply log the error and continue. This makes testing re-try logic harder. 

To address this and ensure the bot performs as expected the tests use an invalid EMP contract that causes the `ExpiringMultiPartyClient` to error out.

**Issue(s)**

Second implementation for issue #1812. Subsequent PR will implement the same logic for the disputer bots.

### CODE REVIEW

File: monitors/SyntheticPegMonitor.js [hunk 1]
```diff
@@ -54,7 +54,7 @@ class SyntheticPegMonitor {
         // `volatilityWindow`: Length of time (in seconds) to snapshot volatility.
         value: uniswapPriceFeed || medianizerPriceFeed ? 60 * 10 : 0, // 10 minutes
         isValid: x => {
-          return typeof x === "number" && x && x >= 0;
+          return typeof x === "number" && x >= 0;
```
Comment (chrismaree) [inline monitors/SyntheticPegMonitor.js:57]: Unrelated to main PR changes. Having this first check previously prevented the `volatilityWindow` from taking on a value of 0.

File: monitors/index.js [hunk 2]
```diff
@@ -151,47 +156,70 @@ async function run(
...
+          await contractMonitor.checkForNewDisputeSettlementEvents();
+          // 1.e Check for new sponsor positions created
+          await contractMonitor.checkForNewSponsors();
 
-      // 3.  Position Collateralization Ratio monitor
-      // 3.a Update the client
-      await empClient.update();
-      // 3.b Check for positions below their CR
-      await crMonitor.checkWalletCrRatio();
+          // 2.  Wallet Balance monitor
```
Comment (chrismaree) [inline monitors/index.js:176]: not sure why this shows up as a diff.

File: financial-templates-lib/logger/SpyTransport.js [hunk 3]
```diff
@@ -9,6 +10,9 @@ class SpyTransport extends Transport {
   }
 
   async log(info, callback) {
+    // Add an `logInTest` option to help with debugging to bots in tests by printing all logs received by winston.
```
Comment (chrismaree) [inline financial-templates-lib/logger/SpyTransport.js:13]: Unrelated to main PR change. Can run tests now like:

```
truffle test ../monitors/test/index.js --network test logInTest
```
and the tests will log all winston outputs. Very useful in debugging.

File: liquidator/test/index.js [hunk 4]
```diff
@@ -33,7 +33,7 @@ contract("index.js", function(accounts) {
 
   let pollingDelay = 0; // 0 polling delay creates a serverless bot that yields after one full execution.
   let executionRetries = 1;
-  let errorRetriesTimeout = 100; // 100 milliseconds between preforming retries
+  let errorRetriesTimeout = 100; // 100 milliseconds between preforming retries.
```
Comment (chrismaree) [inline liquidator/test/index.js:36]: unrelated nit

File: monitors/test/index.js [hunk 5]
```diff
@@ -106,4 +113,67 @@ contract("index.js", function(accounts) {
       assert.notEqual(spyLogLevel(spy, i), "error");
     }
   });
+
+  it("Correctly re-tries after failed execution loop", async function() {
```
Comment (chrismaree) [inline monitors/test/index.js:117]: this test was a massive pain to get right. Our bots are so robust they never throw errors in the while loop. needed to be pretty creative to get _past_ the initial checks but throw in the while loop!

File: package.json [hunk 6]
```diff
@@ -47,7 +47,7 @@
   },
   "husky": {
     "hooks": {
-      "pre-commit": "echo '🕺  Running eslint on staged files' && lint-staged --allow-empty && echo '🏃‍♂️  Running pretty-quick on staged files' && pretty-quick --staged"
+      "pre-commit": "echo '🕺 Running eslint on staged files' && lint-staged --allow-empty && echo '🏃‍♂️ Running pretty-quick on staged files' && pretty-quick --staged"
```
Comment (chrismaree) [inline package.json:50]: unrelated change. spacing in the logs was a bit broken. previous:
image

after change:
image

File: monitors/test/index.js (hunk 5 above)
Comment (nicholaspai) [reply monitors/test/index.js:117]: Interesting way to test this, so this fails on the first iteration  when it tries to update EMP state right?

Comment (chrismaree) [reply monitors/test/index.js:117]: yup. you can create the EMP client but the bot errors out when the emp state updates as the functions the client calls are not defined in this "invalid" emp.
